tokendito/okta_helpers.py

```diff
@@ -115,22 +125,38 @@ def user_mfa_challenge(headers, primary_auth):
         logging.warning(
             "No MFA provided or provided MFA does not exist. [{}]".format(
                 settings.mfa_method))
-        mfa_index = select_preferred_mfa_index(mfa_options)
+        mfa_index = helpers.select_preferred_mfa_index(mfa_options)
 
     # time to challenge the mfa option
     selected_mfa_option = mfa_options[mfa_index]
     logging.debug("Selected MFA is [{}]".format(selected_mfa_option))
 
     mfa_challenge_url = selected_mfa_option['_links']['verify']['href']
-    payload = prepare_payload(stateToken=primary_auth['stateToken'],
-                              factorType=selected_mfa_option['factorType'],
-                              provider=selected_mfa_option['provider'],
-                              profile=selected_mfa_option['profile'])
-    okta_verify_api_method(mfa_challenge_url, payload, headers)
-    logging.debug("mfa_challenge_url [{}] headers [{}]".format(
+
+    payload = helpers.prepare_payload(stateToken=primary_auth['stateToken'],
+                                      factorType=selected_mfa_option['factorType'],
+                                      provider=selected_mfa_option['provider'],
+                                      profile=selected_mfa_option['profile'])
+    selected_factor = okta_verify_api_method(
+        mfa_challenge_url, payload, headers)
+
+    mfa_provider = selected_factor["_embedded"]["factor"]["provider"].lower()
+    logging.debug("MFA Challenge URL: [{}] headers: {}".format(
         mfa_challenge_url, headers))
-    mfa_verify = user_mfa_options(selected_mfa_option,
-                                  headers, mfa_challenge_url, payload, primary_auth)
+
+    if mfa_provider == "duo":
+        payload, headers, callback_url = duo_helpers.authenticate_duo(
+            selected_factor)
+        okta_verify_api_method(callback_url, payload)
+        payload.pop("id", "sig_response")
+        mfa_verify = okta_verify_api_method(
+            mfa_challenge_url, payload, headers)
+    elif mfa_provider == "okta" or mfa_provider == "google":
```

Comment:
how is FIDO handled here? I'm not sure how yubi keys are represented.
Also, `FIDO:webauthn` should be ignored as an option until supported.